### DISCUSSION TITLE
How to convert analysis csv files back to slp or h5 files?

### DISCUSSION THREAD

**Discussion by amblypatty** [talmolab/sleap#1717] [Help!] [answered]
Hello,

I have exported some SLEAP-tracked videos to analysis CSV files. I have edited the csv files without removing any rows or columns. Now, I would like to convert those files back to slp files that are connected to the original videos. 

Ultimately, from the edited csv files, I want to generate annotated videos or 'Render Video Clip with Instances...'. Is there a direct or indirect way that this can be accomplished with functions from the SLEAP CLI or python interface? I cannot figure out to do so. I would greatly appreciate your help!

**Comment by talmo** [comment]
Hi @amblypatty,

Good question! We don't currently support importing those CSV files back in since they are missing a lot of metadata that we normally have in our SLP files.

If you don't mind providing an example CSV together with the SLP file you exported it from, we can come up with a one-off solution as we investigate better ways to implement this workflow.

You can provide those through a link here or via [this form](https://forms.gle/KebBPdjoHw2WMKmV8).

Thanks!

Talmo

**Comment by amblypatty** [comment]
Hi @talmo,

I have uploaded the requested files in the submission form you provided. Perhaps a solution could come from a json file with metadata like skeleton information and video file names that supplemented creation of slp file? However, I am unsure of what metadata is required to create an slp file. Thanks again for your help.

**Comment by talmo** [comment, marked as answer]
Hey @amblypatty,

Thanks for the ping. [Here's a Colab notebook](https://colab.research.google.com/drive/166gFFxm-R3dH5saEBmktMY8PGDImdpOv?usp=sharing) that does the one-off recreation from CSV by just updating the original SLP file.

Let us know if that works for you!

Cheers,

Talmo

**Reply by amblypatty** [reply]
Hi @talmo,

Woot! It did work for the example I submitted! I will attempt to adapt it to run through all my video files. 

Thank you for your help and the clear script. I'm so excited!

Cheers,
Patrick

**Reply by amblypatty** [reply]
Hi @talmo,

Okay, I ran into an issue that I can't seem to figure out when I tried to make the notebook work for an slp file that has multiple videos. From the following code snippet:

```
    for _, row in edited_csv.iterrows():
        lf = original_slp.find(video=original_slp.videos, frame_idx=row.frame_idx)
        
        print(len(lf))
        assert len(lf) > 0, f"Could not find frame {row.frame_idx} in the original labels."
        lf = lf[0]
```
Where I replaced `original_slp.video` with `original_slp.videos` to account for multiple videos in the slp file, I get the following output and Assertion error:

```
0
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[15], line 9
      6 lf = original_slp.find(video=original_slp.videos, frame_idx=row.frame_idx)
      8 print(len(lf))
----> 9 assert len(lf) > 0, f"Could not find frame {row.frame_idx} in the original labels."
     10 lf = lf[0]
     12 found_inst = False

AssertionError: Could not find frame 0 in the original labels.
```
When I print `edited_csv.iloc[:3,:3]` , I get:

```
     track  frame_idx  instance.score
0  track_0          0        0.791155
1  track_0          1        0.766460
2  track_0          2        0.784306
3  track_0          3        0.772470
4  track_0          4        0.769359
```
So, the csv has the frame_idx. I am then thinking that there is a mistake when trying to use the `Labels.find()` method which requires a `Video` and `frame_idx`. When there are multiple videos in the slp file, is this method being incorrectly used to iterate through each video and pull out the frames? I am not sure why the code could not find the first frame. Could you provide some insight as to the possible reasons I might be seeing this? 

Since I am performing this in sleap-io, I can submit this request there if more appropriate. Please advise.
